リファクタリング

taishi-matsumura opened the pull request:
## 概要
#2 で作成したサンプルプロジェクトは初めての試みも多くかなり荒いのでbrush upする。

## 変更内容
- [x] 型aliasを見直す
- [x] ビットマップヘッダーに関する処理を別クラスに切り出す
- [x] streamを引き回すのをやめる
- [x] メッセージの出力先を表すenumを移動
- [x] Factoryクラスが生の参照を返しているのを、unique_ptrを返すようにする
- [x] BitContainer内で保持するデータ型をstd::bitsetに変更する
- [x] データ取得時にvectorを渡してデータを詰めてもらうのではなく、vectorを返してもらうようにする
- [x] vectorが内部で保持しているbufferの再確保・再配置を極力減らす

## 補足
他にも気付いたことがあればどんどん指摘お願いします。

taishi-matsumura committed:
:recycle: 型aliasを見直し

- 型名に仕様が含まれているものを、単純に符号の有無とビット幅を表す名前に修正
- インデックスを表すものはSize型からIndex型に変更
- std::unique_ptrにUniquePtrというaliasを設定
- std::vectorにVectorというaliasを設定

taishi-matsumura committed:
:recycle: streamを引き回して書き出す方法を修正

BitmapWriterがColor及びBitmapからデータを取得して書き出すよう修正

taishi-matsumura committed:
:recycle: ビットマップヘッダーに関する処理を別クラスに切出し

taishi-matsumura committed:
:recycle: メッセージの出力先についてはFactory内で閉じて扱えるのでWriterから移動

taishi-matsumura committed:
:recycle: FactoryがUniquePtrを返すよう修正

Effective Modern C++ 項目18 を参考にした

taishi-matsumura committed:
:recycle: BitContainer内で保持するデータ型をstd::bitsetに変更

それに伴い、ByteContainerを削除

taishi-matsumura committed:
:recycle: autoの積極利用

taishi-matsumura committed:
:recycle: メンバ変数をrename

taishi-matsumura committed:
:zap: 全てのcppでincludeされるヘッダが増えてきたので、プリコンパイル済みヘッダを作成

taishi-matsumura committed:
:recycle: ユニバーサル参照の利用

taishi-matsumura committed:
:recycle: 値型にemplace_backを使っても意味が無いため修正

taishi-matsumura committed:
:pencil2: 無名namespaceにブロックコメントを追加

taishi-matsumura committed:
:recycle: 要素数はコンストラクタ呼出し時に確定するので、イニシャライザで設定するよう修正

taishi-matsumura committed:
:recycle: 冗長なコンパイルスイッチを削除

taishi-matsumura committed:
:recycle: FileHeaderとInfoHeaderのサイズを定数化

taishi-matsumura committed:
:recycle: メソッド名を変更

taishi-matsumura committed:
:recycle: パディングを含む一行のバイト数を取得できるようgetterを追加

taishi-matsumura committed:
:recycle: パディングサイズの算出方法を修正

taishi-matsumura committed:
:sparkles: UInt8のarrayのaliasを作成

taishi-matsumura committed:
:recycle: ビットマップヘッダ部は固定長なので、外部に提供するデータ型をarrayに変更

taishi-matsumura committed:
:recycle: UInt8Array型の作成に伴い、UIntXXXCollectionをUIntXXXVectorにrename

taishi-matsumura committed:
:sparkles: UniquePtr<UInt8Vector>のaliasを作成

taishi-matsumura committed:
:sparkles: VectorにArrayを連結するtemplate関数を作成

vector::insert()もstd::copy()も、イテレータが表に出てくるのがイヤだったのでwrapした

taishi-matsumura committed:
:recycle: パディング時にvector::push_back()ではなくConcatenate()を用いるよう修正

taishi-matsumura committed:
:sparkles: 空のVectorのUniquePtrを生成するwrapperを追加

- これ以後、UInt8VectorPtr型でバイトデータの受け渡しを行うための布石
- 領域の再確保・再配置を回避するために、生成時点でcapacityを設定しておく

taishi-matsumura committed:
:sparkles: UniquePtrが内包しているVectorとコンテナを結合するwrapperを追加

- 一つ前のコミット同様、UInt8VectorPtr型でバイトデータの受け渡しを行うための布石

taishi-matsumura committed:
:recycle: GetLineData()の仕様を変更

呼び元が用意したバッファを受け取りデータを詰めるのではなく、戻り値でデータを返すよう変更

taishi-matsumura committed:
:recycle: Concatenate()の呼び出し形式を揃える

taishi-matsumura committed:
:recycle: GetData()の仕様を変更

呼び元が用意したバッファを受け取りデータを詰めるのではなく、戻り値でデータを返すよう変更

taishi-matsumura committed:
:recycle: UInt8Vectorに直接連結するコードがなくなったのでwrapperも削除

taishi-matsumura committed:
:zap: 領域の再確保・再配置の抑止

taishi-matsumura committed:
:pencil2: 型aliasの適用漏れ

taishi-matsumura committed:
:sparkles: std::functionのaliasを作成

taishi-matsumura commented:
右辺値参照の利用等、まだ改善の余地はあるかと思いますが、一旦これで完成とさせて下さい。

22af473 〜 5eecc42 はコミットを細かく分けていますが、やりたいことは「データ取得の方法を変更する」ということだけです。
完成形を一気にコミットすると差分が非常に多く、reviewerの負担が大きくなり過ぎると判断したためこのような形にしています。

皆さま、レビューをお願い致します。

Harutaka-Tsujino commented on matsumura/AreciboMessage/MessageWriter/Bitmap/PixelIndices.cpp:
```diff
@@ -6,47 +6,55 @@ namespace arecibo_message {
+// Unnamed Namespace
+// ========================================================================================
+namespace {
+void Repeat(Size loopCount, Function<void(Index)> func) {
```
funcの引数がIndexなのでloopCountの型をIndexに変更したほうが適切に思います。
```suggestion
void Repeat(Index loopCount, Function<void(Index)> func) {
```

taishi-matsumura replied:
<strike>
ここは、
* ループカウンターは(VSが提示してくるfor文のsnippetのように)Sizeが良さそう
* funcの引数には頂点のindexを渡して欲しいのでIndexが良い

というの理由で型の選択について悩んだところです。

ちゃんとやるならループカウンターのSize型を引数のIndex型にconvertすべきかと思うのですがいかがでしょう？
で、明示的にconvertするにしても、Size型とIndex型がどちらもstd::size_tのaliasなので単純にキャスト可能なのですがそれは今回/現在たまたまそうなっているだけでは？という点も気色悪かったので、これらのいずれの選択肢も採用せずそのままSize型のループカウンターをIndex型の引数に放り込んでいます。

どうするのがいいんでしょう？
</strike>

すみません、寝言です。全然別のところの話でした。

Harutaka-Tsujino replied:
size_t(Size)はあらゆる型のオブジェクトの最大サイズということなので、
ループカウンタには基本的にSizeを用いるのがいいのかなと思います。

明示的convertについては、私も悩んだ結果、
今回の場合は配列などを使っていないのと、
コード自体が短いのでループカウンタの型もIndexでも良いと思いましたが、
最近はVSのintelligence等で、同じ型が元になっていることが簡単にわかると思うので、
このままでいいと思います。

同じ型のaliasがそれぞれ別の型に認識される方法がないのかと、
調べているとopacue alias(strong typedef)についての記事があったので貼っておきます。
残念ながらデフォルトの機能としては無いようでしたが...
https://cpplover.blogspot.com/2013/09/c1yopaque-alias.html

taishi-matsumura replied:
『寝言です』が寝言でした。まさにご指摘の部分で悩んでいました。
迷走してすみません。

ここの処理の意味合いですが、 `Repeat()` の呼び出し側としては
* `func` で指定する処理を `scale` 回繰り返してもらいたい。
* `func` の引数に渡される `今何回目なのか` という情報を基に、何らかの処理をしたい。現状、以下の2例。
  * 処理対象の行番号(Index型)を算出したい。
  * `indices` に追加するピクセル番号(Index型)を算出したい。

となります。

他方、呼び出される `Repeat()` 側としては
* `func` で指定される何らかの処理を、 `loopCount` で指定された回数繰り返す。
* 今何回目なのかを `func` に伝えてあげる。それを使って呼び出し側が何をしたいのかは知らない。

ということになります。

従って、
* `func` の引数に渡されるのは `今何回目なのか` なので、引数の型をSize型にする。
* `Repeat()` 呼び出し側で、 `今何回目なのか` をSize型から自分が使いたい型(今回は *たまたま* どちらもIndex型)に変換して利用する。

が良いように思います。

その上で、
### 今回の実装では(偶然ではあるが)SizeとIndexはどちらもstd::size_t型だった。
ということを鑑み、
### 特に明示的に変換する処理は書かず、そのまま利用する。
という理屈でいかがでしょう？

P.S.
> 調べているとopacue alias(strong typedef)についての記事があったので貼っておきます。

出た、本の虫！！
OOP 9Rulesの「primitive型のwrap」とかで使えそうですね。(C++1yで検討中とのことなので、C++20には入ったのかな？)
あと、不透過は opaque ですね。GUIプログラムをやってるとよく見かける単語です。

taishi-matsumura replied:
1d9e215 で対応しました。

taishi-matsumura committed:
:ok_hand: 指摘是正(https://github.com/human-osaka-game-2019/AreciboMessage/pull/5#discussion_r380070791)

taishi-matsumura commented:
> vectorが内部で保持しているbufferの再確保・再配置を極力減らす

については、
```C++
(vectorのヘッダファイルのL.725付近)
    template <class... _Valty>
    pointer _Emplace_reallocate(const pointer _Whereptr, _Valty&&... _Val) {
```
にbreak pointを貼り、reallocate処理が走らないことを確認しましたが、vector以外にも各種コピー処理を減らせる箇所があるかと思いますので、そういう観点からもレビューして頂けると幸いです。

戻り値を(参照・ポインタではなく)値型にしている箇所は、RVOが働くことを期待しています。
右辺値参照やmove処理を用いれば、RVOのような『コンパイラによるoptimization』という不安定なものに頼らずに済むかも知れません。
その辺りもご提案頂けると狂喜乱舞します。

taishi-matsumura committed:
:fire: 不要コードの削除

AreciboMessageクラスがpimplを持っていた頃の名残を削除。
デフォルトデストラクタの明示的な実装が必要だった理由については
Effective Modern C++ 項目22を参照。
